packages/realm-web/src/Fetcher.ts

```diff
@@ -0,0 +1,275 @@
+     */
+    public readonly baseUrl: string;
+    private readonly appId: string;
+    private readonly transport: NetworkTransport;
+    private readonly userContext: UserContext;
+    private location: Promise<AppLocation> | null = null;
+
+    /**
+     * Constructs a base transport, which takes paths (prepended by a base URL) instead of absolute urls.
+     *
+     * @param config A configuration of the fetcher.
+     */
+    constructor({
+        baseUrl,
+        appId,
+        transport,
+        userContext,
+        location = null,
+    }: FetcherConfig) {
+        this.baseUrl = baseUrl;
+        this.appId = appId;
+        this.transport = transport;
+        this.userContext = userContext;
+        this.location = location;
+    }
+
+    clone(config: Partial<FetcherConfig>) {
+        return new Fetcher({
+            baseUrl: this.baseUrl,
+            appId: this.appId,
+            transport: this.transport,
+            userContext: this.userContext,
+            location: this.location,
+            ...config,
+        });
+    }
+
+    /**
+     * Fetch a network resource as an authenticated user.
+     *
+     * @param request The request to issue towards the server.
+     * @param retries How many times was this request retried?
+     * @returns A response from requesting with authentication.
+     */
+    public async fetch<RequestBody = any>(
+        request: AuthenticatedRequest<RequestBody>,
+        retries = 0,
+    ): Promise<FetchResponse> {
+        // console.log("Fetching", request.url);
+        const {
+            tokenType = "access",
+            user = this.userContext.currentUser,
+            headers,
+            ...rest
+        } = request;
+
+        try {
+            // Awaiting to intercept errors being thrown
+            const response = await this.transport.fetch({
+                ...rest,
+                headers: {
+                    ...this.buildAuthorizationHeader(user, tokenType),
+                    ...headers,
+                },
+            });
+            // Throw an error if response is not OK
+            const contentType = response.headers.get("content-type");
+            if (!response.ok && contentType?.startsWith("application/json")) {
+                throw new MongoDBRealmError(
+                    request.method,
+                    request.url,
+                    response.status,
+                    response.statusText,
+                    await response.json(),
+                );
+            }
+            // Return the raw response
+            return response;
+        } catch (err) {
+            if (
+                user &&
+                tokenType === "access" &&
+                retries === 0 &&
+                err instanceof MongoDBRealmError &&
+                err.statusCode === 401
+            ) {
+                // Refresh the access token
+                await user.refreshAccessToken();
```

Comment:
Not quite sure I understand this comment.
The user which generated the 401 had an expired access token and we should refresh that regardless if the current user of the app changed in the meantime.

Reply:
Refreshing the access token isn't the interesting bit. The point was more about which user should we use when retrying the request

Reply:
I believe the user which was current when fetch was called should be the user retrying the request. And you're right, there's a race here.
I think a solution is to pass a copy of the request with the user field set explicitly to avoid it defaulting to another currentUser in the retry:

On https://github.com/realm/realm-js/blob/b89519c770a0d10680d94e4df00ea86dbc272005/packages/realm-web/src/Fetcher.ts#L171 should be

```typescript
return this.fetch({ ...request, user }, retries + 1);
```

Reply:
Either that or just add a `for(let retries = 0; retries <= 1; retries++)` loop around the fetch, with a `continue` after `refresh()`. The loop is more what I'm used to seeing, but either way seems fine.